### PR TITLE
Limit 'discarding sample' logs to 1 every 10s

### LINKED ISSUE #2177: Too many warnings logged about remote storage queue being full
**What did you do?**

Ran a Prometheus instance configured to use remote storage on a system with a broken DNS configuration.

I don't have the exact command-line on me. I can construct a likely equivalent if necessary.

**What did you expect to see?**

Error messages in the logs informing me that the remote host couldn't be discovered due to DNS lookup failures.

**What did you see instead? Under which circumstances?**

```
$ cat logs | cut -f 1 -d ' ' --complement | sort | uniq -c | sort -n | grep -v -e '^\s*1\b'
      5 level=info msg="Checkpointing in-memory metrics and chunks..." source="persistence.go:549"
    198 level=warning msg="error sending 100 samples to remote storage: Post http://hostname:80/api/prom/push: dial tcp: lookup hostname on 10.96.0.10:53: dial udp 10.96.0.10:53: i/o timeout" source="queue_manager.go:246"
    369 level=warning msg="error sending 100 samples to remote storage: context deadline exceeded" source="queue_manager.go:246"
6660847 level=warning msg="Remote storage queue full, discarding sample." source="queue_manager.go:169"
```

That is, there were 30000 times as many `Remote storage queue full, discarding sample` messages as there were `dial tcp: lookup hostname on 10.96.0.10:53: dial udp 10.96.0.10:53: i/o timeout` failures.

This lead to the problem being misreported to me ("it says the queue is full") and made it harder than necessary to diagnose the root cause (i.e. that DNS was broken).

**Environment**

Ubuntu 16.04 LTS images running on EC2. Prometheus itself was running as a container in a Kubernetes cluster.

* System information:

Alas, system has been wiped out now.

* Prometheus version: 

Alas, I didn't capture exact output, but it was reported to me as 1.3.1.

* Prometheus configuration file:

Can provide equivalent on request.

* Logs:

All I have left is the summary above.

### PR DESCRIPTION
**Pull request by tomwilkie**
This is the simplest thing I could come up with, but we should probably make something more generic and put it in common/log.  WDYT?

Fixes #2177

### CODE REVIEW

File: vendor/vendor.json [hunk 1]
```diff
@@ -868,6 +868,12 @@
 			"revisionTime": "2016-09-30T00:14:02Z"
 		},
 		{
+			"checksumSHA1": "eFQDEix/mGnhwnFu/Hq63zMfrX8=",
```
Comment (juliusv) [inline vendor/vendor.json]: The `vendor.json` change is here, but the vendored library isn't?

Comment (tomwilkie) [inline vendor/vendor.json]: Bah.

**Comment by juliusv** [comment]
Didn't Google have this log function that both rate-limits, but also tells you how many of the same message were suppressed? Would be cool to do something similar, but this is ok for now IMO. It might just confuse users who are expecting an error to be logged for every failure, without the log message indicating that some of the messages will be suppressed.

**Comment by brian-brazil** [comment]
We also have it as a metric, so it's not so bad.

**Comment by juliusv** [comment]
Yeah, that almost made me think we should drop the log message altogether. But that would be inconsistent with other places (like the notifier queue), where we log too.

So :+1: to this @brian-brazil?

**Comment by tomwilkie** [comment]
> Didn't Google have this log function that both rate-limits, but also tells you how many of the same message were suppressed? Would be cool to do something similar

I started doing that (with a atomic counter of how many were dropped) but decide it was too complicated.  Am happy to give it a go through - can do it as a callback based thing, or as something that implements the log interface.  WDYT?

**Comment by juliusv** [comment]
@tomwilkie Yeah, the tricky bit is to both accumulate, but then definitely summarize the accumulation at some point (even if no further samples have been dropped, but not enough dropped samples have been accumulated yet), which requires a timer. That sounds like overkill. I would just add something like this to the log message:

```go
log.Warn("Remote storage queue full, discarding sample. Multiple subsequent messages of this kind may be suppressed.")
```

**Comment by brian-brazil** [comment]
:+1: 

I think the log messages are useful, as most users wouldn't have alerts for that metrics and it's definitely an error state.

**Comment by juliusv** [comment]
:-1:

**Comment by juliusv** [comment]
Whoops, wrong-sided thumb. I meant :+1:

**Pull request merged**